src/language-js/printer-estree.js

```diff
@@ -1514,6 +1515,27 @@ function printPathNoParens(path, options, print, args) {
         const needsForcedTrailingComma =
           canHaveTrailingComma && lastElem === null;
 
+        const parent = path.getParentNode();
+        // Break array expressions in parameters of jest each function
+        // call expressions.
+        //
+        // test.only.each([
+        //   [1, 1, 2],
+        //   [1, 2, 3],
+        //   [2, 1, 3]
+        // ])(".add(%i, %i)", (a, b, expected) => {
+        //   expect(a + b).toBe(expected);
+        // });
+        const shouldBreak =
+          isJestEachFunctionCall(parent) &&
+          n.elements &&
+          n.elements.length &&
+          hasNewlineInRange(
```

Comment:
Why only if there’s already a newline?

Reply:
I'm worried about the implementation... I think that we should avoid that output depends on the input, but I don't know if we should break an ArrayExpression like this:
```js
// Input
test.each([1, 2, 3])("test", a => {
    // check something
});

// Output
test.each([
  1,
  2,
  3
])("test", a => {
    // check something
});
```
I just came up with it, but it might be better to break only when the ArrayExpression is inside an Array or Object.

Reply:
I think we should explore getting rid of this `hasNewlineInRange` check since that’s something that we try to avoid in Prettier!

Maybe if all items of the array are object literals – only then we do the "table formatting"?

Reply:
> I think we should explore getting rid of this hasNewlineInRange check since that’s something that we try to avoid in Prettier!

Okay, I agree!

> Maybe if all items of the array are object literals – only then we do the "table formatting"?

almost yes. but I think that even in case of all items of the array are **array** literals, "table formatting" should be done.
```js
test.only.each([
  [1, 1, 2],
  [1, 2, 3],
  [2, 1, 3]
])(".add(%i, %i)", (a, b, expected) => {
  expect(a + b).toBe(expected);
});
```

Reply:
Cool, I didn’t know you could have an array of arrays. I agree, we should look for that pattern too.

Reply:
This sounds like a good idea for all array literals, not only for these Jest tests.

Reply:
Yeah, maybe! That could potentially resolve #2716

Reply:
Indeed! Then we don't need to treat the jest each function call specially.

Reply:
I tried to implement it like that on my local, but some test cases were failed as follows.... I think it is not good to ignore ( and force update )this fails. it ’s better to at least listen to user opnion.
```js
Summary of all failing tests
 FAIL  tests/flow/new_react/jsfmt.spec.js
  ● propTypes.js

    expect(value).toMatchSnapshot()

    Received value does not match stored snapshot "propTypes.js 1".

    - Snapshot
    + Received

    @@ -38,11 +38,16 @@
            })
          ).isRequired
        }
      });
      
    - <C statistics={[{}, { label: "", value: undefined }]} />; // error (label is required, value not required)
    + <C
    +   statistics={[
    +     {},
    +     { label: "", value: undefined }
    +   ]}
    + />; // error (label is required, value not required)
      
      var props: Array<{ label: string, value?: number }> = [
        {},
        { label: "", value: undefined }
      ]; // error (same as ^)

      88 |           )
      89 |         )
    > 90 |       ).toMatchSnapshot();
         |         ^
      91 |     });
      92 | 
      93 |     for (const parser of parsers.slice(1)) {

      at Object.test (tests_config/run_spec.js:90:9)

 FAIL  tests/flow/annot/jsfmt.spec.js
  ● annot.js

    expect(value).toMatchSnapshot()

    Received value does not match stored snapshot "annot.js 1".

    - Snapshot
    + Received

    @@ -83,19 +83,25 @@
      var arr: Array<number> = [1, 2, "..."];
      
      // array sugar
      var array: number[] = [1, 2, "..."];
      
    - var matrix: number[][] = [[1, 2], [3, 4]];
    + var matrix: number[][] = [
    +   [1, 2],
    +   [3, 4]
    + ];
      var matrix_parens: number[][] = matrix;
      
      var nullable_array: ?(number[]) = null;
      var nullable_array_parens: ?(number[]) = nullable_array;
      
      var array_of_nullable: (?number)[] = [null, 3];
      
    - var array_of_tuple: [number, string][] = [[0, "foo"], [1, "bar"]];
    + var array_of_tuple: [number, string][] = [
    +   [0, "foo"],
    +   [1, "bar"]
    + ];
      var array_of_tuple_parens: [number, string][] = array_of_tuple;
      
      type ObjType = { "bar-foo": string, "foo-bar": number };
      var test_obj: ObjType = { "bar-foo": "23" };
      

      88 |           )
      89 |         )
    > 90 |       ).toMatchSnapshot();
         |         ^
      91 |     });
      92 | 
      93 |     for (const parser of parsers.slice(1)) {

      at Object.test (tests_config/run_spec.js:90:9)

 FAIL  tests/flow/object_assign/jsfmt.spec.js
  ● apply.js

    expect(value).toMatchSnapshot()

    Received value does not match stored snapshot "apply.js 1".

    - Snapshot
    + Received

    @@ -8,11 +8,12 @@
      (Object.assign.apply(null, [{}, {a: 1}, {b: 'foo'}]): {a: number, b: string});
      
      =====================================output=====================================
      // @flow
      
    - (Object.assign.apply(null, [{}, { a: 1 }, { b: "foo" }]): {
    -   a: number,
    -   b: string
    - });
    + (Object.assign.apply(null, [
    +   {},
    +   { a: 1 },
    +   { b: "foo" }
    + ]): { a: number, b: string });
      
      ================================================================================

      88 |           )
      89 |         )
    > 90 |       ).toMatchSnapshot();
         |         ^
      91 |     });
      92 | 
      93 |     for (const parser of parsers.slice(1)) {

      at Object.test (tests_config/run_spec.js:90:9)

  ● spread.js

    expect(value).toMatchSnapshot()

    Received value does not match stored snapshot "spread.js 1".

    - Snapshot
    + Received

    @@ -28,12 +28,15 @@
      
      (Object.assign({}, ...arrOfObjs): { foo: number }); // Error: string ~> number
      (Object.assign({}, ...roArrOfObjs): { foo: number }); // Error: string ~> number
      (Object.assign({}, ...tup): { foo: string, bar: boolean }); // Error: number ~> boolean
      
    - (Object.assign({}, ...[{ a: 1 }, { b: "foo" }], ...[{ c: true }]): {
    -   a: number,
    -   b: true,
    -   c: boolean
    - }); // Error: 'foo' => true
    + (Object.assign(
    +   {},
    +   ...[
    +     { a: 1 },
    +     { b: "foo" }
    +   ],
    +   ...[{ c: true }]
    + ): { a: number, b: true, c: boolean }); // Error: 'foo' => true
      
      ================================================================================

      88 |           )
      89 |         )
    > 90 |       ).toMatchSnapshot();
         |         ^
      91 |     });
      92 | 
      93 |     for (const parser of parsers.slice(1)) {

      at Object.test (tests_config/run_spec.js:90:9)

 FAIL  tests/flow/arraylib/jsfmt.spec.js
  ● array_lib.js

    expect(value).toMatchSnapshot()

    Received value does not match stored snapshot "array_lib.js 1".

    - Snapshot
    + Received

    @@ -103,11 +103,15 @@
      
        var total = [0, 1, 2, 3].reduce(function(a, b) {
          return a + b;
        });
      
    -   var flattened = [[0, 1], [2, 3], [4, 5]].reduce(function(a, b) {
    +   var flattened = [
    +     [0, 1],
    +     [2, 3],
    +     [4, 5]
    +   ].reduce(function(a, b) {
          return a.concat(b);
        });
      
        /* Added later, because the above is insufficient */
      

      88 |           )
      89 |         )
    > 90 |       ).toMatchSnapshot();
         |         ^
      91 |     });
      92 | 
      93 |     for (const parser of parsers.slice(1)) {

      at Object.test (tests_config/run_spec.js:90:9)
```

Reply:
This is the only change that I don't like:
```diff
- <C statistics={[{}, { label: "", value: undefined }]} />; // error (label is required, value not required)
+ <C
+   statistics={[
+     {},
+     { label: "", value: undefined }
+   ]}
+ />; // error (label is required, value not required)
```
My proposal then is to break if all the elements are array or object literals with more than 1 entry.

```js
// won't be broken
[[], [1, 2, 3]];
[[1], [1]];
[{ a }, {}];
[{ a, b, c}, { a }]

// will be broken
[[1, 2], [1, 2, 3]];
[[1, 0], [1, 0]];
[{ a, b }, { c, d }];
[{ a, b, c}, { a, b }]
```

Reply:
I like the `statistics` case broken on several lines 🤷‍♂ 

But @thorn0’s latest suggestion sounds worth exploring.

Reply:
/cc @lydell @thorn0
I pushed commits to implement thorn0's suggestion, and add some snapshot tests. Please review!